### ISSUE TITLE
Target property is not matched / Remove existing shortcut file

### ISSUE THREAD

**Issue by ericcsinger** [mkht/DSCR_Shortcut#11]
Hi,
Thank you for putting this together.  Seeing something a little strange.  I have a simple shortcut need that keeps failing the test process of DSC.

The DSC works, it creates the proper shortcut and the shortcut works, but it always fails the test with the following error "Target property is not matched"

I'm only specifying the following:

cShortcut "cShortcut_$($ConfigurationBaseName)_shortcut"

    {
    Path = "C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Microsoft Log Parser Studio.lnk" 
    Target = "C:\Blah\IT\Infrastructure Services\SysAdmins\DSC\Download Cache\Microsoft\LogParser Studio\2.0.0.100\LPS.exe"
    }

Any idea what i might be doing wrong?  Like I said,the shortcut does work and it does exist, it just fails the test process.

**Comment by mkht** [comment]
@ericcsinger Thanks for reporting issue. And sorry for the late reply.  🙏 

Probably, the existence shortcut file `Microsoft Log Parser Studio.lnk` has some properties that has not been specified in the DSC resource.

In DSCR_Shortcut v1.3.7, We test all properties of the shortcut file including those not explicitly specified.
We fixed this issue in v1.3.8. Please try to use the new version.

**Comment by ericcsinger** [comment]
Thanks for your reply, just gave it another go, no luck.  Here is the verbose output.

``
VERBOSE: [-0701]: LCM:  [ Start  Resource ]  [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut]
VERBOSE: -0701]: LCM:  [ Start  Test     ]  [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut]
VERBOSE: [-0701]:                            [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut] Trying to get file properties from "C:\ProgramData\Microsoft\Windows\Star
t Menu\Programs\Microsoft Log Parser Studio.lnk"
VERBOSE: [-0701]:                            [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut] Target property is not matched!
VERBOSE: [-0701]:                            [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut] Test returns False
VERBOSE: [-0701]: LCM:  [ End    Test     ]  [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut] Fall
``

Any additional info i can provide to help troubleshoot?  Thank you for putting this DSC resource together btw, appreciate it.

**Comment by mkht** [comment]
@ericcsinger  That's too bad that you did not solve the issue...

I see the verbose messages. 
This line indicates that the ***target*** path of the shortcut file does not match the path specified in the configuration.
```
VERBOSE: [-0701]: [[cShortcut]cShortcut_Microsoft Log Parser 2.0.0.100_shortcut] Target property is not matched!
```

I've tried invoking the below configuration.
But, In my environment, I could not reproduce the issue.
The target file has been created successfully, and `Test-DscConfiguration` returns TRUE.
```PowerShell
# Test for reproduce
Configuration test
{
    Import-DscResource -ModuleName DSCR_Shortcut -ModuleVersion 1.3.8

    cShortcut "cShortcut_Microsoft Log Parser 2.0.0.100_shortcut"
    {
        Path   = "C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Microsoft Log Parser Studio.lnk" 
        Target = "C:\Blah\IT\Infrastructure Services\SysAdmins\DSC\Download Cache\Microsoft\LogParser Studio\2.0.0.100\LPS.exe"
    }
}
``` 

----
Try following the steps below to investigate the cause in more detail.

1. Please check the properties of the created shortcut file.
  In my test, it has the following properties: (This is exactly what I specified in the configuration)
  ![image](https://user-images.githubusercontent.com/25482642/55960998-e94e6180-5ca8-11e9-87e8-9d518de849e1.png)

1. If possible, zip and attach the created shortcut file.

1. Please paste all configuration code you are using.

**Comment by ericcsinger** [comment]
Let me know if you need additional info

[LogParser.zip](https://github.com/mkht/DSCR_Shortcut/files/3069648/LogParser.zip)

**Comment by mkht** [comment]
@ericcsinger OK, I understand the cause of the issue.

Here is the result of expanding all of your configuration parameters:
(For the first two, I put a temporary path.)
```PowerShell
Name                           Value                                                                                        
----                           -----                                                                                        
DSCDownloadCacheRootDirectory  C:\DSCDownloadCacheRootDirectory                                                             
DSCFileshareRootDirectory      C:\DSCFileshareRootDirectory                                                                 
ZipFileName1                   LPS.zip                                                                                      
ZipDestinationFolder1          C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100                        
ZipDestinationFilePath1        C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100\LPS.zip                
ZipSourceFolderPath1           C:\DSCFileshareRootDirectory\Microsoft\LogParser Studio\2.0.0.100                            
ZipSourceFilePath1             C:\DSCFileshareRootDirectory\Microsoft\LogParser Studio\2.0.0.100\LPS.zip                    
InstallerFileName1             LPS.exe                                                                                      
DestinationFolder1             C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100\                       
DestinationFilePath1           C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100\\LPS.exe               
DesiredVersion1                2.0.0.100                                                                                    
ConfigurationBaseName          Microsoft Log Parser 2.0.0.100                                                               
ShortcutName                   Microsoft Log Parser Studio.lnk                                                              
ShortcutFolderPath             C:\ProgramData\Microsoft\Windows\Start Menu\Programs                                         
ShortcutFilePath               C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Microsoft Log Parser Studio.lnk         
ShortCutTarget                 C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100\\LPS.exe  
```

It is worth noting that path delimiters of the `ShortcutTarget`  are doubled.
```
C:\DSCDownloadCacheRootDirectory\Microsoft\LogParser Studio\2.0.0.100\\LPS.exe 
                                                                    ^^^^^
```

When the DSCR_Shortcut module creates a shortcut file, path delimiters are parsed to single `\`. 
As a result, The `Target` property of the shortcut file to be:
`"C:\DownloadCache\Microsoft\LogParser Studio\2.0.0.100\LPS.exe"` 

The test method compares two parameters that differ only by this path delimiter, it is determined that the parameters do NOT match.

It is easy to avoid this issue by changing the processing on the module side.
But, if anything, you have to essentially correct such misuse of path delimiters.

**Comment by mib82** [comment]
I agree with @ericcsinger. There is a bug in **Test** mechanism that is checking Target and it's not matched when there comes to:
* **fileshare** path:
\\\\file-server\fileshare
* **web** path:
http:// or https://

**Comment by mkht** [comment]
Hi @mib82 
Thank you for reporting the issue.
I have recognized that the issue occurs some times. But I unfortunately couldn't reproduce it.
Can you provide me the steps of reproduce?

**Comment by mib82** [comment]
Hi @mkht 

- DSC used in Azure Automation Account 
- as node Virtual Machine with Windows Server 2019
- cShortcut -> Target: fileshare or web path

**Comment by mkht** [comment]
Hi @mib82 
Sorry for the delay in getting back to you.

I still haven't been able to reproduce the problem with the shared file shortcuts. Can you provide the configuration you are using and a shortcut file that can reproduce the problem?

We have identified the cause of the issue with the web shortcuts and fixed it. I will upload it to the PowerShell gallery soon.

**Comment by mkht** [comment]
I have released [v2.2.0](https://www.powershellgallery.com/packages/DSCR_Shortcut/2.2.0).
This release includes a fix for an issue where the tests for web path shortcuts did not work properly.

The issue of tests failing for shared files has not been fixed as I could not reproduce it, but I am closing this Issue. Please feel free to open a new Issue if needed.
